src/platform/windows/mod.rs

```diff
@@ -0,0 +1,1486 @@
+                    }
+                }
+
+                // We may have already been closed, or the read resulted in us being closed.
+                // If so, add that to the result and remove the reader from our list.
+                if reader.closed {
+                    win32_trace!("[# {:?}] receiver {:?} ({}) -- now closed!", *self.iocp, *reader.handle, reader.set_id.unwrap());
+                    selection_results.push(OsIpcSelectionResult::ChannelClosed(reader.set_id.unwrap()));
+                    remove_index = Some(reader_index);
+                } else {
+                    try!(reader.start_read());
+                }
+            }
+
+            if remove_index.is_some() {
+                self.readers.swap_remove(remove_index.unwrap());
+            }
+
+            // if we didn't dequeue at least one complete message -- we need to loop through GetQueuedCS again;
+            // otherwise we're done.
+            if !selection_results.is_empty() {
+                break;
+            }
+        }
+
+        win32_trace!("select() -> {} results", selection_results.len());
+        Ok(selection_results)
+    }
+}
+
+impl OsIpcSelectionResult {
+    pub fn unwrap(self) -> (u64, Vec<u8>, Vec<OsOpaqueIpcChannel>, Vec<OsIpcSharedMemory>) {
+        match self {
+            OsIpcSelectionResult::DataReceived(id, data, channels, shared_memory_regions) => {
+                (id, data, channels, shared_memory_regions)
+            }
+            OsIpcSelectionResult::ChannelClosed(id) => {
+                panic!("OsIpcSelectionResult::unwrap(): receiver ID {} was closed!", id)
+            }
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct OsIpcSharedMemory {
+    handle: WinHandle,
+    ptr: *mut u8,
+    length: usize,
+}
+
+unsafe impl Send for OsIpcSharedMemory {}
+unsafe impl Sync for OsIpcSharedMemory {}
+
+impl Drop for OsIpcSharedMemory {
+    fn drop(&mut self) {
+        unsafe {
+            kernel32::UnmapViewOfFile(self.ptr as LPVOID);
+        }
+    }
+}
+
+impl Clone for OsIpcSharedMemory {
+    fn clone(&self) -> OsIpcSharedMemory {
+        unsafe {
+            let mut handle = dup_handle(&self.handle).unwrap();
+            OsIpcSharedMemory::from_handle(handle.take(), self.length).unwrap()
+        }
+    }
+}
+
+impl PartialEq for OsIpcSharedMemory {
+    fn eq(&self, other: &OsIpcSharedMemory) -> bool {
+        self.handle == other.handle
+    }
+}
+
+impl Deref for OsIpcSharedMemory {
+    type Target = [u8];
+
+    #[inline]
+    fn deref(&self) -> &[u8] {
+        assert!(!self.ptr.is_null() && self.handle.is_valid());
+        unsafe {
+            slice::from_raw_parts(self.ptr, self.length)
+        }
+    }
+}
+
+impl OsIpcSharedMemory {
+    #[allow(exceeding_bitshifts)]
+    fn new(length: usize) -> Result<OsIpcSharedMemory,WinError> {
+        unsafe {
+            assert!(length < u32::max_value() as usize);
+            let (lhigh, llow) = (0 as u32, (length & 0xffffffffusize) as u32);
```

Comment:
I've read up on the exceeding bitshifts situation (ugh); and I see now that we do indeed need some kind of explicit handling here -- sorry for this.

Still, I think the original approach with `#[allow(exceeding_bitshifts)]` was pretty fragile and confusing. I'd vote for more explicit conditional compilation, like this:

```rust
#[cfg(target_pointer_width = "64")]
let (lhigh, llow) = ((length >> 32) as u32, (length & 0xffffffff) as u32);
#[cfg(target_pointer_width = "32")]
let (lhigh, llow) = (0, length as u32);
```

As an additional advantage, if we ever encounter a different pointer size, this will fail at compile time, rather than producing something unexpected. (Might be a hypothetical case -- but better safe than sorry...)

Yet another option is using `checked_shr()`:

```rust
let (lhigh, llow) = (length.checked_shr(32).unwrap_or(0) as u32, (length & 0xffffffff) as u32);
```

This seems the simplest and cleanest option by far. It might result in a branch being performed at run time, unless the compiler is smart enough to elide it -- but the overhead should be insignificant (especially since it's a predictable branch); and I'd say the simplification justifies it...